renku/core/management/workflow/providers/toil.py

```diff
@@ -0,0 +1,244 @@
+            directory_content = dict()
+            for f in location.rglob("*"):
+                directory_content[str(f.relative_to(basedir))] = import_function(str(f))
+            file_locations[p.actual_value] = directory_content
+        else:
+            file_locations[p.actual_value] = import_function(str(location))
+
+    return file_locations
+
+
+def importFileWrapper(storage: AbstractFileStore, file_path: str) -> FileID:
+    """Wrap importFile accept file:// URIs."""
+    file_uri = file_path if ":/" in file_path else f"file://{file_path}"
+    return storage.importFile(file_uri)
+
+
+def process_children(parent: Job, children: List[Plan], dag: nx.DiGraph, basedir: Path, storage: AbstractFileStore):
+    """Recursively process children of a workflow."""
+
+    outputs = list()
+    import_function = functools.partial(importFileWrapper, storage)
+    for child in children:
+        file_metadata = _upload_files(import_function, child.inputs, basedir)
+        childJob = SubprocessToilJob(child, file_metadata, parent_promise=parent.rv())
+        outputs.append(parent.addFollowOn(childJob).rv())
+        outputs += process_children(childJob, nx.neighbors(dag, child), dag, basedir, storage)
+    return outputs
+
+
+def initialize_jobs(job, children, basedir, dag):
+    """Creates the Toil execution plan for the given workflow DAG."""
+
+    job.fileStore.logToMaster("executing renku DAG")
+    outputs = list()
+    import_function = functools.partial(importFileWrapper, job.fileStore)
+    for workflow in children:
+        file_metadata = _upload_files(import_function, workflow.inputs, basedir)
+        childJob = SubprocessToilJob(workflow, file_metadata)
+        outputs.append(job.addChild(childJob).rv())
+        outputs += process_children(childJob, nx.neighbors(dag, workflow), dag, basedir, job.fileStore)
+
+    return outputs
+
+
+class ToilProvider(IWorkflowProvider):
+    """A workflow executor provider using toil."""
+
+    @hookimpl
+    def workflow_provider(self):
+        """Workflow provider name."""
+        return (self, "toil")
+
+    @hookimpl
+    def workflow_execute(self, dag: nx.DiGraph, basedir: Path, config: Dict[str, Any]):
+        """Executes a given workflow DAG using Toil."""
+        tmpdir = (Path(RENKU_HOME) / RENKU_TMP).resolve()
+        if not tmpdir.exists():
+            tmpdir.mkdir()
+        options = Job.Runner.getDefaultOptions(str(tmpdir / uuid.uuid4().hex))
+        options.logLevel = "ERROR"
+        options.clean = "always"
+
+        if config:
+            [setattr(options, k, v) for k, v in config.items()]
+
+        outputs = list()
+        try:
+            with Toil(options) as toil:
+                start_wfs = next(nx.topological_generations(dag))
+                rootJob = Job.wrapJobFn(initialize_jobs, start_wfs, basedir, dag)
+
+                job_outputs = toil.start(rootJob)
+
+                num_outputs = functools.reduce(lambda a, b: a + len(b.values()), job_outputs, 0)
```

Comment:
```suggestion
                num_outputs = sum(map(len, job_outputs))
```
?

Reply:
does `map(..., job_outputs)` iterates through the values of `job_outputs` instead of the `(k,v)` tuple?

Reply:
oh right, I missed the `.values()`. then `num_outputs = sum(map(lambda x: len(x.values()), job_outputs))`

in any case, I think using `sum()` makes the intention more clear.